### PR TITLE
Removed tags; Simplified interceptor code; Added logging fields editability.

### PR DESCRIPTION
**Pull request by bwplotka**
Kind of biggy, but should be much cleaner.

The only surprise is that tracing is using `logging.Fields` to both fill span metadata (trace ID, sampled or not, span ID) as well. I think this is useful AND it was like that before 🤗 

At least it's a clean separation of responsibilities who is using what vs generic tags.

Fixes https://github.com/grpc-ecosystem/go-grpc-middleware/issues/382

NOTE:  I removed tracing package completely since in new PR we complete rewrite it anyway.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by bwplotka** [comment]
image

**Comment by bwplotka** [comment]
No idea @yashrsharma44 how you PR passed linting... it fails now 
image ;p

File: go.sum [hunk 1]
```diff
@@ -30,6 +30,7 @@ github.com/google/go-cmp v0.3.1/go.mod h1:8QqcDgzrUqlUb/G2PQTWiueGozuR1884gddMyw
 github.com/google/go-cmp v0.4.0/go.mod h1:v8dTdLbMG2kIc/vJvl+f65V22dbkXbowE6jgT/gNBxE=
 github.com/google/go-cmp v0.5.0 h1:/QaMHBdZ26BB3SSst0Iwl10Epc+xhTquomWX0oZEB6w=
 github.com/google/go-cmp v0.5.0/go.mod h1:v8dTdLbMG2kIc/vJvl+f65V22dbkXbowE6jgT/gNBxE=
+github.com/grpc-ecosystem/go-grpc-middleware v1.2.2 h1:FlFbCRLd5Jr4iYXZufAvgWN6Ao0JrI5chLINnUXDDr0=
```
Comment (johanbrandhorst) [inline go.sum:33]: I assume this is because of an incorrect import somewhere? Should this be here?

File: interceptors/auth/examples_test.go [hunk 2]
```diff
@@ -9,8 +9,9 @@ import (
 	pb "google.golang.org/grpc/examples/helloworld/helloworld"
 	"google.golang.org/grpc/status"
 
+	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/logging"
+
 	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/auth"
```
Comment (johanbrandhorst) [inline interceptors/auth/examples_test.go:14]: Nit: import grouping

File: interceptors/logging/doc.go [hunk 3]
```diff
@@ -5,19 +5,19 @@
 /*
 logging is a "parent" package for gRPC logging middlewares.
 
-The gRPC logging middleware populates request-scoped data to `grpc_ctxtags.Tags` that relate to the current gRPC call
-(e.g. service and method names).
+The gRPC logging middleware populates request-scoped data to `logging.Fields` that relate to the current gRPC call
+(e.g. service and method names). You can laverage that data using `logging.ExtractFields` and `logging.InjectFields`.
```
Comment (johanbrandhorst) [inline interceptors/logging/doc.go:9]: Spelling
```suggestion
(e.g. service and method names). You can leverage that data using `logging.ExtractFields` and `logging.InjectFields`.
```

File: interceptors/logging/logging.go [hunk 4]
```diff
@@ -37,46 +35,110 @@ var (
...
+		ServiceFieldKey, c.Service,
+		MethodFieldKey, c.Method,
+		MethodTypeFieldKey, string(c.Typ),
 	}
 }
 
-// Fields represents logging fields. It has to have even number of elements (pairs).
-type Fields []string
+func (f Fields) Iter() FieldsIter {
+	return iter{i: -2, f: f}
```
Comment (johanbrandhorst) [inline interceptors/logging/logging.go:59]: Plese add a comment as to why this is -2.

File: interceptors/logging/logging.go [hunk 5]
```diff
@@ -37,46 +35,110 @@ var (
...
+}
 
-// ErrorToCode function determines the error code of an error
-// This makes using custom errors with grpc middleware easier
-type ErrorToCode func(err error) codes.Code
+type FieldsIter interface {
+	Next() bool
+	At() (k, v string)
+}
```
Comment (johanbrandhorst) [inline interceptors/logging/logging.go:65]: What is this interface used for? Can't we just return a `map[string]string`? This seems needlessly complicated.

File: interceptors/logging/logging.go [hunk 6]
```diff
@@ -37,46 +35,110 @@ var (
...
+	if i.i+1 == len(i.f) {
+		// Non even number of elements, add empty string.
+		return i.f[i.i], ""
+	}
+	return i.f[i.i], i.f[i.i+1]
 }
 
-// Decider function defines rules for suppressing any interceptor logs
-type Decider func(fullMethodName string, err error) Decision
+// AppendUnique returns fields which is the union of all keys with the added values having lower priority.
```
Comment (johanbrandhorst) [inline interceptors/logging/logging.go:93]: This text confused me, but after reading the source, I think I'd reword this like this:
```suggestion
// AppendUnique returns fields which is the union of all keys. Any keys that already exist in the log fields will take precedence over duplicates in add.
```

File: interceptors/logging/logging.go [hunk 7]
```diff
@@ -37,46 +35,110 @@ var (
...
+	}
+	return n
 }
 
-// ServerPayloadLoggingDecider is a user-provided function for deciding whether to log the server-side
-// request/response payloads
-type ServerPayloadLoggingDecider func(ctx context.Context, fullMethodName string, servingObject interface{}) bool
+// ExtractFields returns logging.Fields object from the Context.
+// Logging interceptor adds fields into context when used.
+// If no one injected fields before ExtractFields returns empty Fields.
```
Comment (johanbrandhorst) [inline interceptors/logging/logging.go:122]:
```suggestion
// If there are no fields in the context, returns an empty Fields value.
```

File: interceptors/logging/logging.go [hunk 8]
```diff
@@ -37,46 +35,110 @@ var (
...
+	}
+	n := make(Fields, len(t))
+	copy(n, t)
+	return n
+}
 
-// ClientPayloadLoggingDecider is a user-provided function for deciding whether to log the client-side
-// request/response payloads
-type ClientPayloadLoggingDecider func(ctx context.Context, fullMethodName string) bool
+// InjectFields allows to add logging.Fields that will be used logging interceptor in the path of given context (if any).
```
Comment (johanbrandhorst) [inline interceptors/logging/logging.go:135]:
```suggestion
// InjectFields allows adding Fields to any existing Fields that will be used by the logging interceptor.
```

File: interceptors/logging/logging.go [hunk 9]
```diff
@@ -37,46 +35,110 @@ var (
...
+func InjectFields(ctx context.Context, f Fields) context.Context {
+	return context.WithValue(ctx, fieldsCtxMarkerKey, ExtractFields(ctx).AppendUnique(f))
+}
 
-// JsonPbMarshaller is a marshaller that serializes protobuf messages.
-type JsonPbMarshaler interface {
+// JsonPBMarshaller is a marshaller that serializes protobuf messages.
+type JsonPBMarshaller interface {
```
Comment (johanbrandhorst) [inline interceptors/logging/logging.go:141]: The US English spelling of this is `Marshaler`. This naming is consistent with the gRPC gateway (https://pkg.go.dev/github.com/grpc-ecosystem/grpc-gateway/v2/runtime#JSONPb), but note that Go protobuf APIv2 calls this `protojson`.
```suggestion
// JsonPBMarshaler is a marshaler that serializes protobuf messages.
type JsonPBMarshaler interface {
```

File: interceptors/tracing/interceptors_test.go [hunk 10]
```diff
@@ -21,14 +21,15 @@ import (
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 
-	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/tags"
+	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/logging"
+
```
Comment (johanbrandhorst) [inline interceptors/tracing/interceptors_test.go:25]: Import grouping

File: interceptors/tracing/server.go [hunk 11]
```diff
@@ -13,28 +14,28 @@ import (
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/grpclog"
 
+	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/logging"
+
```
Comment (johanbrandhorst) [inline interceptors/tracing/server.go:18]: Import grouping.

**Comment by yashrsharma44** [comment]
> No idea @yashrsharma44 how you PR passed linting... it fails now on this PR here 🤔
> 
> image ;p

I have added a PR - https://github.com/grpc-ecosystem/go-grpc-middleware/pull/395 to fix the same, let me know if anything else is required.

**Comment by bwplotka** [comment]
We need to get back to this (note to my self)

**Comment by olivierlemasle** [comment]
@bwplotka Do you still want to work on this?

**Comment by bwplotka** [comment]
Well I want, but I don't mind someone else taking this if they have time. 🤗  I am going for some vacation this week, and will be back in August, so won't touch it earlier then this ): 

Help wanted! I will have time review/comment

**Comment by bwplotka** [comment]
on it!

File: interceptors/logging/logging.go (hunk 5 above)
Comment (bwplotka) [reply interceptors/logging/logging.go:65]: This is because kit logging way of specifying fields - we incorporated that but using our "safer" implementation of using fields encoded in []string

**Comment by bwplotka** [comment]
PTAL @yashrsharma44 - addressed comments & rebased. Thanks for detailed review @johanbrandhorst !

**Comment by bwplotka** [comment]
NOTE:  I removed tracing package completely since in new PR we complete rewrite it anyway.

**Comment by bwplotka** [comment]
PTAL @yashrsharma44

File: interceptors/logging/logging.go [hunk 12]
```diff
@@ -37,34 +35,89 @@ var (
...
+		MethodFieldKey, c.Method,
+		MethodTypeFieldKey, string(c.Typ),
 	}
 }
 
-// Fields represents logging fields. It has to have even number of elements (pairs).
-type Fields []string
+// Iter returns FieldsIterator.
+func (f Fields) Iter() FieldsIterator {
+	// We start from -2 as we iterate over two items per iteration and first iteration will advance iterator to 0.
```
Comment (yashrsharma44) [inline interceptors/logging/logging.go:60]: Should we raise an error in case the length of `Fields` array is not even? We always expect the key-value format, so we can raise an error. Thoughts?

File: interceptors/logging/doc.go [hunk 13]
```diff
@@ -2,21 +2,21 @@
 // Licensed under the Apache License 2.0.
 
 /*
-logging is a "parent" package for gRPC logging middlewares.
+Package logging is a "parent" package for gRPC logging middlewares.
 
-The gRPC logging middleware populates request-scoped data to `grpc_ctxtags.Tags` that relate to the current gRPC call
-(e.g. service and method names).
+The gRPC logging middleware populates request-scoped data to `logging.Fields` that relate to the current gRPC call
+(e.g. service and method names). You can leverage that data using `logging.ExtractFields` and `logging.InjectFields`.
```
Comment (yashrsharma44) [inline interceptors/logging/doc.go:8]:
```suggestion
(e.g. service and method names). You can extract/inject data in the propagated context using `logging.ExtractFields` and `logging.InjectFields`.
```

File: providers/kit/examples_test.go [hunk 14]
```diff
@@ -30,11 +30,9 @@ func Example_initializationWithCustomLevels() {
 	// Create a server, make sure we put the tags context before everything else.
 	_ = grpc.NewServer(
 		grpc.ChainUnaryInterceptor(
-			tags.UnaryServerInterceptor(),
 			logging.UnaryServerInterceptor(kit.InterceptorLogger(logger), opts...),
 		),
 		grpc.ChainStreamInterceptor(
```
Comment (yashrsharma44) [inline providers/kit/examples_test.go:35]: No more tags interceptor for inserting key-values in the context, that is a superb cleanup of code!

File: interceptors/reporter.go [hunk 15]
```diff
@@ -55,16 +55,34 @@ func SplitMethodName(fullMethod string) (string, string) {
 	return "unknown", "unknown"
 }
 
-func FullMethod(service, method string) string {
-	return fmt.Sprintf("/%s/%s", service, method)
+type CallMeta struct {
```
Comment (yashrsharma44) [inline interceptors/reporter.go:58]: Nice use of Meta for abstracting Client/Server implementation 💪

File: interceptors/skip/interceptor_test.go [hunk 16]
```diff
@@ -29,8 +28,13 @@ const (
 	keyMethod   = "skip.method"
 )
 
+var ctxKey = &ctxKeMarker{}
+
+type ctxKeMarker struct{}
```
Comment (yashrsharma44) [inline interceptors/skip/interceptor_test.go:33]:
```suggestion
type ctxKeyMarker struct{}
```
Is it intentional?

**Review by yashrsharma44 — APPROVED** [review]
Some nitpicks/questions, otherwise
Looks awesome to me! 🥇

**Comment by bwplotka** [comment]
Thanks for comments Yash, will try to get them sorted soon

**Comment by bwplotka** [comment]
Addressed 👍🏽 Thanks for review!
